File: leetcode/Python/0340-h-longest-substring-with-at-most-k-distinct-characters.py
"""
Time: O(n)
Space: O(1)
"""
import collections
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        cHash = collections.defaultdict(int)
        j = 0
        maxLen = 0
        for i in range(len(s)):
            while len(cHash) <= k and j < len(s):
                cHash[s[j]] += 1
                j += 1
                if len(cHash) <= k:
                    maxLen = max(maxLen, j - i)
            cHash[s[i]] -= 1
            if cHash[s[i]] == 0:
                del cHash[s[i]]
        return maxLen

File: leetcode/C++/0340-h-longest-substring-with-at-most-k-distinct-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        int n = s.length();
        vector<int> letters(256, 0);
        int i = 0, j = 0;
        int count = 0, res = 0;
        while (i <= j && j < n) {
            while (count <= k && j < n) {
                if (letters[(int)s[j]] == 0) {
                    if (count == k)
                        break;
                    count++;
                }
                letters[(int)s[j]]++;
                j++;
            }
            res = max(res, j - i);
            if (letters[(int)s[i]] == 1)
                count--;
            letters[(int)s[i]]--;
            i++;
        }
        return res;
    }
};